File: taurus-protect-sdk-python/taurus_protect/services/currency_service.py
"""Currency service for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, List, Optional

from taurus_protect.errors import NotFoundError
from taurus_protect.mappers.currency import currencies_from_dto, currency_from_dto
from taurus_protect.models.currency import Currency
from taurus_protect.services._base import BaseService

if TYPE_CHECKING:
    from taurus_protect._internal.openapi.api.currencies_api import CurrenciesApi


class CurrencyService(BaseService):
    """
    Service for retrieving currency information.

    Provides methods to list all currencies, get specific currencies,
    and retrieve the base currency configured for the tenant.

    Example:
        >>> # List all enabled currencies
        >>> currencies = client.currencies.list()
        >>> for currency in currencies:
        ...     print(f"{currency.symbol}: {currency.name}")
        >>>
        >>> # Get currency by blockchain and network
        >>> eth = client.currencies.get_by_blockchain("ETH", "mainnet")
        >>> print(f"{eth.name} has {eth.decimals} decimals")
        >>>
        >>> # Get the base currency (e.g., USD, EUR, CHF)
        >>> base = client.currencies.get_base_currency()
        >>> print(f"Base currency: {base.symbol}")
    """

    def __init__(
        self,
        api_client: Any,
        currencies_api: "CurrenciesApi",
    ) -> None:
        """
        Initialize the currency service.

        Args:
            api_client: The OpenAPI client instance.
            currencies_api: The CurrenciesApi instance from OpenAPI client.
        """
        super().__init__(api_client)
        self._api = currencies_api

    def list(
        self,
        show_disabled: bool = False,
        include_logo: bool = False,
    ) -> List[Currency]:
        """
        Get all currencies.

        Args:
            show_disabled: If True, includes currencies disabled by business rules.
                Defaults to False (only enabled currencies).
            include_logo: If True, includes logo URLs in the response.
                Defaults to False (logos omitted for performance).

        Returns:
            List of currencies.

        Raises:
            APIError: If API request fails.
        """
        try:
            resp = self._api.wallet_service_get_currencies(
                show_disabled=show_disabled if show_disabled else None,
                include_logo=include_logo if include_logo else None,
            )

            result = getattr(resp, "currencies", None) or getattr(resp, "result", None)
            return currencies_from_dto(result) if result else []
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, APIError):
                raise
            raise self._handle_error(e) from e

    def get(self, currency_id: str) -> Currency:
        """
        Get a currency by ID.

        This method retrieves currency details using the currency ID.
        For looking up by blockchain and network, use `get_by_blockchain()`.

        Args:
            currency_id: The unique currency identifier.

        Returns:
            The currency.

        Raises:
            ValueError: If currency_id is empty.
            NotFoundError: If currency is not found.
            APIError: If API request fails.
        """
        self._validate_required(currency_id, "currency_id")

        # Get all currencies and filter by ID
        # The API doesn't have a direct get-by-id endpoint
        try:
            currencies = self.list(show_disabled=True)
            for currency in currencies:
                if currency.id == currency_id:
                    return currency

            raise NotFoundError(f"Currency with id '{currency_id}' not found")
        except NotFoundError:
            raise
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def get_by_blockchain(
        self,
        blockchain: str,
        network: str,
        contract_address: Optional[str] = None,
        token_id: Optional[str] = None,
    ) -> Currency:
        """
        Get a currency by blockchain and network.

        This retrieves the native currency for a blockchain/network pair,
        or a specific token if contract_address is provided.

        Args:
            blockchain: Blockchain type (e.g., "ETH", "BTC").
            network: Network name (e.g., "mainnet", "testnet").
            contract_address: Token contract address. If not provided,
                returns the native currency.
            token_id: For blockchains with multi-asset contracts (e.g., ALGO, XTZ),
                specifies which token within the contract.

        Returns:
            The currency.

        Raises:
            ValueError: If required arguments are missing.
            NotFoundError: If currency is not found.
            APIError: If API request fails.
        """
        self._validate_required(blockchain, "blockchain")
        self._validate_required(network, "network")

        try:
            resp = self._api.wallet_service_get_currency(
                unique_currency_filter_blockchain=blockchain,
                unique_currency_filter_network=network,
                unique_currency_filter_token_contract_address=contract_address,
                unique_currency_filter_token_id=token_id,
                show_disabled=True,
                currency_id=None,
                include_logo=None,
            )

            result = getattr(resp, "currency", None) or getattr(resp, "result", None)
            currency = currency_from_dto(result)

            if currency is None:
                raise NotFoundError(
                    f"Currency for blockchain '{blockchain}' network '{network}' not found"
                )

            return currency
        except NotFoundError:
            raise
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, (APIError, ValueError)):
                raise
            raise self._handle_error(e) from e

    def get_base_currency(self) -> Currency:
        """
        Get the base currency configured for the tenant.

        The base currency is used for fiat valuations and is typically
        configured as CHF, EUR, or USD.

        Returns:
            The base currency.

        Raises:
            NotFoundError: If no base currency is configured.
            APIError: If API request fails.
        """
        try:
            resp = self._api.wallet_service_get_base_currency()

            result = getattr(resp, "currency", None) or getattr(resp, "result", None)
            currency = currency_from_dto(result)

            if currency is None:
                raise NotFoundError("Base currency not configured")

            return currency
        except NotFoundError:
            raise
        except Exception as e:
            from taurus_protect.errors import APIError

            if isinstance(e, APIError):
                raise
            raise self._handle_error(e) from e

File: taurus-protect-sdk-typescript/src/services/currency-service.ts
/**
 * Currency service for Taurus-PROTECT SDK.
 *
 * Provides methods for retrieving currency information.
 */

import { NotFoundError, ValidationError } from '../errors';
import type { CurrenciesApi } from '../internal/openapi/apis/CurrenciesApi';
import { currenciesFromDto, currencyFromDto } from '../mappers/currency';
import type {
  Currency,
  GetCurrencyByBlockchainOptions,
  ListCurrenciesOptions,
} from '../models/currency';
import { BaseService } from './base';

/**
 * Service for retrieving currency information.
 *
 * Provides methods to list all currencies, get specific currencies,
 * and retrieve the base currency configured for the tenant.
 *
 * @example
 * ```typescript
 * // List all enabled currencies
 * const currencies = await currencyService.list();
 * for (const currency of currencies) {
 *   console.log(`${currency.symbol}: ${currency.name}`);
 * }
 *
 * // Get currency by blockchain and network
 * const eth = await currencyService.getByBlockchain({
 *   blockchain: 'ETH',
 *   network: 'mainnet',
 * });
 * console.log(`${eth.name} has ${eth.decimals} decimals`);
 *
 * // Get the base currency (e.g., USD, EUR, CHF)
 * const base = await currencyService.getBaseCurrency();
 * console.log(`Base currency: ${base.symbol}`);
 * ```
 */
export class CurrencyService extends BaseService {
  private readonly currenciesApi: CurrenciesApi;

  /**
   * Creates a new CurrencyService instance.
   *
   * @param currenciesApi - The CurrenciesApi instance from the OpenAPI client
   */
  constructor(currenciesApi: CurrenciesApi) {
    super();
    this.currenciesApi = currenciesApi;
  }

  /**
   * Lists all currencies.
   *
   * @param options - Optional filtering options
   * @returns Array of currencies
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // List only enabled currencies
   * const currencies = await currencyService.list();
   *
   * // List all currencies including disabled ones
   * const allCurrencies = await currencyService.list({ showDisabled: true });
   *
   * // Include logo URLs
   * const withLogos = await currencyService.list({ includeLogo: true });
   * ```
   */
  async list(options?: ListCurrenciesOptions): Promise<Currency[]> {
    return this.execute(async () => {
      const response = await this.currenciesApi.walletServiceGetCurrencies({
        showDisabled: options?.showDisabled,
        includeLogo: options?.includeLogo,
      });

      const result =
        (response as Record<string, unknown>).currencies ??
        (response as Record<string, unknown>).result;
      return currenciesFromDto(result as unknown[]);
    });
  }

  /**
   * Gets a currency by ID.
   *
   * This method retrieves currency details using the currency ID.
   * For looking up by blockchain and network, use `getByBlockchain()`.
   *
   * @param currencyId - The unique currency identifier
   * @returns The currency
   * @throws {@link ValidationError} If currencyId is empty
   * @throws {@link NotFoundError} If currency is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const currency = await currencyService.get('eth-mainnet');
   * console.log(`Symbol: ${currency.symbol}, Decimals: ${currency.decimals}`);
   * ```
   */
  async get(currencyId: string): Promise<Currency> {
    if (!currencyId || currencyId.trim() === '') {
      throw new ValidationError('currencyId is required');
    }

    return this.execute(async () => {
      // Get all currencies and filter by ID since API doesn't have a direct get-by-id endpoint
      const currencies = await this.list({ showDisabled: true });
      const currency = currencies.find((c) => c.id === currencyId);

      if (!currency) {
        throw new NotFoundError(`Currency with id '${currencyId}' not found`);
      }

      return currency;
    });
  }

  /**
   * Gets a currency by blockchain and network.
   *
   * This retrieves the native currency for a blockchain/network pair,
   * or a specific token if contractAddress is provided.
   *
   * @param options - Blockchain lookup options
   * @returns The currency
   * @throws {@link ValidationError} If required arguments are missing
   * @throws {@link NotFoundError} If currency is not found
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * // Get native currency
   * const eth = await currencyService.getByBlockchain({
   *   blockchain: 'ETH',
   *   network: 'mainnet',
   * });
   *
   * // Get a specific token
   * const usdc = await currencyService.getByBlockchain({
   *   blockchain: 'ETH',
   *   network: 'mainnet',
   *   contractAddress: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
   * });
   * ```
   */
  async getByBlockchain(options: GetCurrencyByBlockchainOptions): Promise<Currency> {
    if (!options.blockchain || options.blockchain.trim() === '') {
      throw new ValidationError('blockchain is required');
    }
    if (!options.network || options.network.trim() === '') {
      throw new ValidationError('network is required');
    }

    return this.execute(async () => {
      const response = await this.currenciesApi.walletServiceGetCurrency({
        uniqueCurrencyFilterBlockchain: options.blockchain,
        uniqueCurrencyFilterNetwork: options.network,
        uniqueCurrencyFilterTokenContractAddress: options.contractAddress,
        uniqueCurrencyFilterTokenID: options.tokenId,
        showDisabled: true,
      });

      const result =
        (response as Record<string, unknown>).currency ??
        (response as Record<string, unknown>).result;
      const currency = currencyFromDto(result);

      if (!currency) {
        throw new NotFoundError(
          `Currency for blockchain '${options.blockchain}' network '${options.network}' not found`
        );
      }

      return currency;
    });
  }

  /**
   * Gets the base currency configured for the tenant.
   *
   * The base currency is used for fiat valuations and is typically
   * configured as CHF, EUR, or USD.
   *
   * @returns The base currency
   * @throws {@link NotFoundError} If no base currency is configured
   * @throws {@link APIError} If API request fails
   *
   * @example
   * ```typescript
   * const baseCurrency = await currencyService.getBaseCurrency();
   * console.log(`Base currency: ${baseCurrency.symbol}`);
   * ```
   */
  async getBaseCurrency(): Promise<Currency> {
    return this.execute(async () => {
      const response = await this.currenciesApi.walletServiceGetBaseCurrency();

      const result =
        (response as Record<string, unknown>).currency ??
        (response as Record<string, unknown>).result;
      const currency = currencyFromDto(result);

      if (!currency) {
        throw new NotFoundError('Base currency not configured');
      }

      return currency;
    });
  }
}
